keps/sig-storage/20180515-svcacct-token-volumes.md

```diff
@@ -159,6 +162,40 @@ sources:
     audience: ca.istio.io
 ```
 
+### File Permission
+
+
+1. If there is only one UserID (non-root) shared by all the containers in a pod,
+set the owner of projected service account volume to the UserID and mode to 600.
+2. Otherwise, set the owners to be the UserID of the first container in the pod
+and the group either specified by `fsGroup` in the `PodSecurityContext` or
+unused GroupID and set the mode to 0640.
+
+UserID can be obtained by the following three ways:
+1. `runAsUser` in `SecurityContext`
+2. fetch UserID from image status
```

Comment:
are all images pulled and present by the time we need to mount the files and choose a file mode? can the preferred uid/gid for an image change over the life of a pod if a container image is changed in the pod spec or is restarted/repulled?

Reply:
it is indeterminate because the volume manager and kuberuntime manager are in different go routines but we can set up a channel.

good point. the uid/gid will not change if a new image with different uid/gid is pulled. this will be an issue for uid but not for gid because the user in the new image will be added to the gid. to work around this changing uid issue, we have to change the ownership each pulling, which requires implementing a SetOwnership method without remounting. talked to @saad-ali about the implications of this method: 1. "the operation is recursive and can take many many minutes to complete for volumes with lots of files and directories." 2. some files might be used by containers while being updated with new permissions. 

the complexity of solving the second issue is high. ideally we should push container runtime side add a functionality to set ownership of volume mounts. see https://github.com/moby/moby/issues/2259.

or we pursue the second alternative.

Reply:
another way to get around the image issue: if the new image has a different uid, we add it to the pre-installed gid.

Reply:
I don't think the recursive SetOwnership is a concern for this particularly case. It's only problematic for very large volumes with lots of files. But changing permissions of files while they're in use is a concern.